osquery/core/query.cpp

```diff
@@ -167,27 +167,43 @@ Status Query::addNewResults(QueryData current_qd,
   return Status(0, "OK");
 }
 
-Status serializeRow(const Row& r,
+class DocAppenderVisitor : public boost::static_visitor<> {
+ public:
+  DocAppenderVisitor(JSON& dc, rj::Value& ob) : doc(dc), obj(ob) {}
+  void operator()(const std::string& key, const int64_t& i) const {
+    doc.add(key, i, obj);
+  }
+
+  void operator()(const std::string& key, const std::string& str) const {
+    doc.addRef(key, str, obj);
+  }
+
+  void operator()(const std::string& key, const double& d) const {
+    doc.add(key, d, obj);
+  }
+
+ private:
+  JSON& doc;
+  rj::Value& obj;
+};
+
+Status serializeRow(const RowTyped& r,
                     const ColumnNames& cols,
                     JSON& doc,
                     rj::Value& obj) {
-  if (cols.empty()) {
-    for (const auto& i : r) {
-      doc.addRef(i.first, i.second, obj);
-    }
-  } else {
-    for (const auto& c : cols) {
-      auto i = r.find(c);
-      if (i != r.end()) {
-        doc.addRef(c, i->second, obj);
-      }
+  DocAppenderVisitor const visitor(doc, obj);
+
+  for (const auto& i : r) {
+    if (cols.empty() ||
+        std::find(cols.begin(), cols.end(), i.first) != cols.end()) {
+      boost::variant<std::string> key = i.first;
+      boost::apply_visitor(visitor, key, i.second);
```

Comment:
`DocAppenderVisitor` seems to be taking references and storing them on the json object. Aren't you creating and storing references to stack objects here when creating and passing `key`?